### PR TITLE
pycbc_make_skymap: add support for generating skymaps from simulated data & injections

### PR DESCRIPTION
**Pull request by ArthurTolley**
Continuation of:  #3640 
Had to reset my fork due to messing things up (a common occurence).

Previous suggestions from @titodalcanton :

- Do not tie together the injection file and fake data, i.e. let the user add an injection to real or simulated data.
- Instead of hardcoding the PSD models, create a new multidetector command-line option that works like this: --fake-strain H1:aLIGOZeroDetSuperPowers L1:aLIGOZeroDetSuperPowers V1:AdvVirgoAmazingSensitivity K1:KAGRAUnconceivablyGreatPSD (have a look at similar options for examples, such as --detector-status). The value for each detector will have to be picked out and sent to the corresponding --fake-strain option passed to pycbc_single_template.
- You will also want the ability to specify the random seeds for the simulated strain, so create an additional multidetector option like this: --fake-strain-seed H1:1234 L1:2345 V1:3456. Again, each value will have to be sent to the pycbc_single_template call corresponding to the right detector.
- Other --frame-strain-* options can probably be omitted, and if pycbc_single_template wants them explicitly, it is probably fine to hardcode them to something sensible (but only if --fake-strain has been given).

This commit looks at adding multi-ifo options for `--fake-strain` and `--fake-strain-seed`.
These are placeholder names as I'm not clued up on the standard in pycbc. I don't think I can call it `--fake-strain` as it might be confusing? But it would look a lot cleaner.

On the point of not tying the injection file and fake data together, I need to do some testing with `pycbc_single_template` to see how this would work.

### CODE REVIEW

File: bin/pycbc_make_skymap [hunk 1]
```diff
@@ -448,6 +462,14 @@ if __name__ == '__main__':
...
                         help='Lists of local frame files, e.g., '
                              'H1:/path/to/frame/file L1:/path/to/frame/file')
+    parser.add_argument('--injection-file', type=str,
+                        help='Optional path to an injection file.')
+    parser.add_argument('--multiifo-fake-strain', nargs='+', type=str,
+                        action=MultiDetOptionAppendAction, metavar="IFO:FAKE_STRAIN",
+                        help="The set of fake-strains for each detector")
+    parser.add_argument('--multiifo-fake-strain-seed', nargs='+', type=str,
+                        action=MultiDetOptionAppendAction, metavar="IFO:FAKE_STRAIN_SEED",
+                        help="The set of fake-strains-seeds for each detector")
```
Comment (titodalcanton) [inline bin/pycbc_make_skymap:472]: The "multiifo" part in the option names are not needed and may create confusion. Just `--fake-strain` and `--fake-strain-seed` should be fine (basically matching the existing options in the strain module).

File: bin/pycbc_make_skymap [hunk 2]
```diff
@@ -465,4 +487,4 @@ if __name__ == '__main__':
          test_event=not opt.enable_production_gracedb_upload,
          custom_frame_files=opt.custom_frame_file,
          approximant=opt.approximant, detector_state=opt.detector_state,
-         veto_definer=opt.veto_definer)
+         veto_definer=opt.veto_definer, injection_file=opt.injection_file)
```
Comment (titodalcanton) [inline bin/pycbc_make_skymap:490]: You also need to pass the values given to `--fake-strain` and `--fake-strain-seed`.

**Comment by ArthurTolley** [comment]
Hi @titodalcanton I believe I have implemented the changes you requested

File: bin/pycbc_make_skymap [hunk 3]
```diff
@@ -224,11 +240,12 @@ def main(trig_time, mass1, mass2, spin1z, spin2z, f_low, f_upper, sample_rate,
         command.append(tmpdir+"/PSD_"+str(rough_time)+"_"+ifo+".txt")
         command.append("--output-file")
         command.append(tmpdir+"/SNRTS_"+str(rough_time)+"_"+ifo+".hdf")
-
+        
         log_path = tmpdir + '/pycbc_single_template_{}_{}_log.txt'.format(str(rough_time), ifo)
         log_file = open(log_path, 'w')
         log_file.write(' '.join(command) + '\n\n')
         log_file.flush()
+        #print(command)
```
Comment (titodalcanton) [inline bin/pycbc_make_skymap:248]: Please remove this.

File: bin/pycbc_make_skymap [hunk 4]
```diff
@@ -224,11 +240,12 @@ def main(trig_time, mass1, mass2, spin1z, spin2z, f_low, f_upper, sample_rate,
         command.append(tmpdir+"/PSD_"+str(rough_time)+"_"+ifo+".txt")
         command.append("--output-file")
         command.append(tmpdir+"/SNRTS_"+str(rough_time)+"_"+ifo+".hdf")
-
+
```
Comment (titodalcanton) [inline bin/pycbc_make_skymap:243]: Please revert this change.

File: bin/pycbc_make_skymap [hunk 5]
```diff
@@ -448,6 +465,14 @@ if __name__ == '__main__':
                         action=MultiDetOptionAppendAction,
                         help='Lists of local frame files, e.g., '
                              'H1:/path/to/frame/file L1:/path/to/frame/file')
+    parser.add_argument('--injection-file', type=str,
+                        help='Optional path to an injection file.')
+    parser.add_argument('--fake-strain', nargs='+', type=str,
+                        action=MultiDetOptionAppendAction, metavar="IFO:FAKE_STRAIN",
+                        help="The set of fake-strains for each detector")
+    parser.add_argument('--fake-strain-seed', nargs='+', type=str,
+                        action=MultiDetOptionAppendAction, metavar="IFO:FAKE_STRAIN_SEED",
```
Comment (titodalcanton) [inline bin/pycbc_make_skymap:474]: The seed is an `int`, not a `str`. Also, can you try to remove the `nargs` kwarg? I think it is no longer needed now (you may have to rebase this PR on master though).

File: bin/pycbc_make_skymap [hunk 6]
```diff
@@ -184,38 +189,49 @@ def main(trig_time, mass1, mass2, spin1z, spin2z, f_low, f_upper, sample_rate,
...
-                    msg = 'Channel name in {} is not {}'.format(
-                            custom_frame, channel_names[ifo])
+        
+        if injection_file:
+            command.append("--injection-file"),command.append(injection_file)
+        
+        if fake_strain:
+            command.append("--fake-strain")
+            command.append(fake_strain[ifo][0])
+            if fake_strain_seed:
```
Comment (titodalcanton) [inline bin/pycbc_make_skymap:199]: This should be
```python
if fake_strain_seed is not None:
```
One may want to use a random seed of zero.

File: bin/pycbc_make_skymap [hunk 7]
```diff
@@ -168,7 +172,8 @@ def main(trig_time, mass1, mass2, spin1z, spin2z, f_low, f_upper, sample_rate,
         coinc_results['foreground/'+ifo+'/window'] = window
         coinc_results['foreground/'+ifo+'/sample_rate'] = int(sample_rate)
         coinc_results['foreground/'+ifo+'/template_id'] = i
-
+        
+
```
Comment (titodalcanton) [inline bin/pycbc_make_skymap:176]: Please do not add this extra space.

File: bin/pycbc_make_skymap [hunk 8]
```diff
@@ -27,6 +27,8 @@ from ligo.gracedb.rest import GraceDb
 def default_frame_type(time, ifo):
     """Sensible defaults for frame types based on interferometer and time.
     """
+    if opt.fake_strain:
+        return ifo + '_FAKE_DATA'
```
Comment (titodalcanton) [inline bin/pycbc_make_skymap:31]: I forget if we discussed this already; could you explain why this is needed? If one wants fake strain, specifying the frame type should not be necessary (i.e. this function should not even be called). Same comment for `default_channel_name()` below.

Comment (spxiwh) [reply bin/pycbc_make_skymap:31]: The frame_type indeed is not needed. The channel name is needed (weirdly) because it is what is used to figure out the ifo. I don't like that that is the case, but I think this is not the place to change that.

File: bin/pycbc_make_skymap (hunk 5 above)
Comment (ArthurTolley) [reply bin/pycbc_make_skymap:470]: Has this changed?

The `fake-strain-seed` option is fed into `pycbc_single_template` in the form of 'H1:1234 L1:5678'. Is this not a string (the individual H1:1234)?

When I remove `nargs='+'` I am presented the error: `The character ':' is used to distinguish the ifo and the value. It must be given exactly once for all entries`.

I believe I have rebased correctly but I might have slipped up somewhere.

File: bin/pycbc_make_skymap (hunk 4 above)
Comment (ArthurTolley) [reply bin/pycbc_make_skymap:240]: I seem to have rewritten my previous commits by force-pushing my changes when trying to rebase. 

Even before that, I was unsure what you meant with your comment?

Comment (titodalcanton) [reply bin/pycbc_make_skymap:240]: I meant to avoid introducing those extra white spaces (there should be no spaces in blank lines).

Comment (ArthurTolley) [reply bin/pycbc_make_skymap:240]: Ah, I see now.
Yes, this is something I am very poor at keeping track of!

File: bin/pycbc_make_skymap (hunk 5 above)
Comment (titodalcanton) [reply bin/pycbc_make_skymap:470]: Hmm I see what you mean. You could indeed literally pass the strings `H1:1234` and `L1:5678` to `pycbc_single_template`, but note that those are two separate strings, which should be passed as two separate arguments, i.e. not a single string with a white space. If you want to go this way, then you should include `nargs='+'`, but *remove* `action=MultiDetOptionAppendAction` (since you will just ask argparse to simply pass the list of strings, instead of interpreting the H1:1234 syntax).

Comment (ArthurTolley) [reply bin/pycbc_make_skymap:470]: When removing `action=MultiDetOptionAppendAction` it stores the `fake_strain` arguments in a list which it then can't loop over for each ifo simply by doing `fake_strain[ifo]`. I can see that a similar argument `--channel-name` creates an empty dictionary at the beginning of the code. However, because `--fake-strain` and `--fake-strain-seed` are optional arguments I don't know if this is what I want to do.

Is there another way to extract the corresponding fake strain values for each ifo?

Comment (titodalcanton) [reply bin/pycbc_make_skymap:470]: Yeah, sorry, I was not thinking straight in my last comment.

The initial code with `type=str` was indeed correct given that you just pass the value as a string. However, `type=int` has the small additional benefit that it makes it clear that the option is supposed to take an integer (as opposed to any random string). When you pass the value to `pycbc_single_template`, you then need to convert it back to a string like this:
```python
command.append(str(fake_strain_seed[ifo][0]))
```

Another comment I should have made before is that the action should probably be `action=MultiDetOptionAction`, and then you can switch from
```python
command.append(str(fake_strain_seed[ifo][0]))
```
to just
```python
command.append(str(fake_strain_seed[ifo]))
```
(the same goes for `--fake-strain`). The difference is that `MultiDetOptionAppendAction` allows you to pass multiple values for each detector, like `H1:123 H1:456 L1:345 L1:789`. This is appropriate for `--custom-frame-files`, but not for `--fake-strain` and `--fake-strain-seed`.

As for `nargs='+'`, it *should* work without it, but you can just leave it in otherwise.

File: bin/pycbc_make_skymap [hunk 9]
```diff
@@ -185,37 +187,48 @@ def main(trig_time, mass1, mass2, spin1z, spin2z, f_low, f_upper, sample_rate,
...
+            command.append("--injection-file"),command.append(injection_file)
+
+        if fake_strain:
+            command.append("--fake-strain")
+            command.append(fake_strain[ifo][0])
+            if fake_strain_seed is not None:
+                command.append("--fake-strain-seed")
+                command.append(fake_strain_seed[ifo][0])
+
+        if not fake_strain:
```
Comment (titodalcanton) [inline bin/pycbc_make_skymap:200]: You can turn this line into an `else:`.

File: bin/pycbc_make_skymap (hunk 5 above)
Comment (titodalcanton) [reply bin/pycbc_make_skymap:470]: The need for `nargs='+'` should have been removed by #3663, by the way.

Comment (ArthurTolley) [reply bin/pycbc_make_skymap:470]: That all makes a lot of sense!

Your explanation of the difference between the `action` arguments has cleared up a few things in my head so thank you for that 😁.

I also now understand what you mean with the `int` and `str` differences when dealing with `--fake-strain-seed`, I will perform these changes now.

Comment (ArthurTolley) [reply bin/pycbc_make_skymap:470]: Without the `nargs='+'` argument for `--fake-strain-seed` I get the error `ValueError: invalid literal for int() with base 10: 'H'`.

File: bin/pycbc_make_skymap [hunk 10]
```diff
@@ -185,37 +188,48 @@ def main(trig_time, mass1, mass2, spin1z, spin2z, f_low, f_upper, sample_rate,
         command.append("--trigger-time"),command.append(str(np.round(trig_time,5)))
         command.append("--window"),command.append(str(window))
 
-        if custom_frame_files is None or ifo not in custom_frame_files:
-            command.append("--frame-type")
-            command.append(frame_types[ifo])
+        if injection_file:
+            command.append("--injection-file"),command.append(injection_file)
+
+        if fake_strain:
```
Comment (titodalcanton) [inline bin/pycbc_make_skymap:194]: I just double-checked a couple things and I think this would be more correct like this:
```python
if fake_strain[ifo] is not None:
```

File: bin/pycbc_make_skymap [hunk 11]
```diff
@@ -185,37 +188,48 @@ def main(trig_time, mass1, mass2, spin1z, spin2z, f_low, f_upper, sample_rate,
...
+            command.append("--fake-strain")
+            command.append(fake_strain[ifo])
+            if fake_strain_seed is not None:
```
Comment (titodalcanton) [inline bin/pycbc_make_skymap:197]: Similar comment as above, I think this should be
```python
if fake_strain_seed[ifo] is not None:
```

File: bin/pycbc_make_skymap [hunk 12]
```diff
@@ -16,7 +16,8 @@ import pycbc
 from pycbc.filter import sigmasq
 from pycbc.io import live, WaveformArray
 from pycbc.types import TimeSeries, FrequencySeries, \
-        MultiDetMultiColonOptionAction, MultiDetOptionAppendAction
+        MultiDetMultiColonOptionAction, MultiDetOptionAppendAction, \
```
Comment (ArthurTolley) [inline bin/pycbc_make_skymap:19]: @titodalcanton (getting to grips with commenting and reviewing, sorry if i've pinged you twice) Should the `MultiDet` options be placed on 3 lines or is this okay?

Comment (titodalcanton) [reply bin/pycbc_make_skymap:19]: I think this is fine. When in doubt about code style, [PEP8](https://www.python.org/dev/peps/pep-0008/) is the reference we try to follow.

**Review by titodalcanton — APPROVED** [review]
I am happy with this now.

@ArthurTolley have you tested that your change does what you want, and that the examples we have in `examples/make_skymap` still run correctly?

**Comment by ArthurTolley** [comment]
@titodalcanton I will be doing a final testing today, I didn't realise those examples existed so that is great!

The examples I am using are:
- Full simulation, simulated strain + injection
- Injecting into data, non-simulated strain + injection
- local frame files
- no simulation at all

I'm hoping this is comprehensive enough.
